### ISSUE TITLE
30% accuracy on single batch size validation

### ISSUE THREAD

**Issue by AhmedEmadEldinHussin** [Woodman718/CapsNets#7]
Im really interested in your work and really liked the result you got 
Actually i have a problem 
When i make the batch size is 1 on the validation dataset the accuracy drops to 33% accuracy 
Do you have any recommendations to solve this issue

**Comment by Woodman718** [comment]
Hyperparameters do have an impact on the model. 
Not only the batch size, but also parameters like nw (number of workers) and lr (learning rate) significantly influence the final results.
You can try the following code to determine the batch size for your task when searching:

---------------------------------------------------
from utils import ImageShow,metrics_scores
network.load_state_dict(torch.load(save_PATH))
dict_bat = {}
for k in range(8,49):
    T_size = k

    print(f"T_size:{k}")
    get_data('test')

    # for j in range(3):
    test()

    if dict_bat.get(k) is None or dict_bat[k] < test_acc:
            dict_bat[k] = test_acc

show.conclusion(img_title=img_title)
print(sorted(dict_bat.items(), key=lambda x: x[1], reverse=True)[0:9])

metrics_scores(test_evl_result,n_classes,cla_dict)

----------------------------------------------------------
The code for metrics_scores and the show function is in the 'utils.py' file.

**Comment by AhmedEmadEldinHussin** [comment]
yes, I mean batch t-size has effect on model performance 
each batch size gives different accuracy and when I make T-size to 1 the accuracy drops down

**Comment by Woodman718** [comment]
"Dynamic routing and K-means are very similar. Data on small batches may not be sufficient to capture the global structure of the data, leading to instability in clustering results. Another potential factor is the loss function, where I use the Nuclear Norm, employing singular values as M+ and M-. The matrix condition number is proportional to its norm. The higher the condition number, the more unstable the matrix, and the loss function will be more sensitive to errors in positive samples."
The definition principles for M+ and M- involve Functional Analysis, and no analysis in this area was found in the works of Hinton and related papers, including my own. It provides significant assistance in enhancing the interpretability of the model. You may consider discussing this with friends who are proficient in Functional Analysis, and there might be fascinating results.
Additionally, a smaller batch size may aid the model in better generalizing to new samples but may also introduce some randomness. To further explore the stability and generalization capabilities of the model, it may be necessary to refactor the model and modify the loss function. For example, in the Primary_Caps class, stacking inputs after eight transformations increased the computational load of the model without achieving the expected improvement in performance.
